### PR TITLE
Framejuly

### PR DESCRIPTION
**Pull request by jmframe**
SRC code with dataset, trainer, validation, etc.

### CODE REVIEW

File: src/neuralngen/dataset/camelsus.py
Comment (FyavariFyavari) [inline src/neuralngen/dataset/camelsus.py:1]: In the following, three comments are brought.
1) 

 Original:
 
 
def __init__(
    self,
    cfg,
    is_train: bool,
    period: str,
    basin: str = None,
    additional_features=[],
    id_to_int={},
    scaler={},
    run_dir=None,
    do_load_scalers=True,
):
    self.additional_features = additional_features
    self.id_to_int = id_to_int
    self.scaler = scaler
    …

COMMENT:
Using None + `… or []/{}`` prevents all instances sharing the same list/dict, eliminating weird cross-instance bugs.


This part could be like below:

def __init__(
    self,
    cfg: Config,
    is_train: bool,
    period: str,
    basin: Optional[str] = None,
    additional_features: Optional[List[str]] = None,
    id_to_int: Optional[Dict[str,int]] = None,
    scaler: Optional[Dict[str,Any]] = None,
    run_dir: Optional[Path] = None,
    do_load_scalers: bool = True,
) -> None:
    # eliminate mutable defaults
    self.additional_features = additional_features or []
    self.id_to_int          = id_to_int or {}
    self.scaler             = scaler or {}
    …


2)

 Original:
 
 
if basins:
    if any(b not in df.index for b in basins):
        raise ValueError("Some basins are missing static attributes.")
    df = df.loc[basins]

COMMENT:


Collecting exactly which IDs are missing makes debugging much faster than a generic “some are missing” message.


This part could be like this:

if basins:
    missing = set(basins) - set(df.index)
    if missing:
        raise ValueError(f"Missing static attributes for basins: {sorted(missing)}")
    df = df.loc[basins]


3)

Original:


if attributes_to_keep:
    for attr in attributes_to_keep:
        if attr in df.columns:
            cols_to_keep.append(attr)
        else:
            raise ValueError(f"Attribute {attr} not found in CAMELS attributes.")
df = df.loc[:, cols_to_keep]


COMMENT:

Pulling-together all checks before slicing ensures we never accidentally drop a required column or select an unintended one.


This part could be like:


required = ['gauge_lat','gauge_lon','area_gages2']
keep = required.copy()


for attr in attributes_to_keep or []:
    if attr not in df.columns:
        raise ValueError(f"Requested attribute '{attr}' not in CAMELS data")
    keep.append(attr)


df = df[keep]

File: src/neuralngen/training/basetrainer.py
Comment (FyavariFyavari) [inline src/neuralngen/training/basetrainer.py:1]: there are few comments in the following:

1)
Original:


class BaseTrainer:
    def __init__(self, cfg, model, dataset_class):
        """
        Parameters
        ----------
        cfg : Config object
            Configuration parameters.
        model : torch.nn.Module
            Your neural network model.
        dataset_class : Dataset class that takes (cfg, period) as arguments.
        """
        …
COMMENT:

Explicit type hints improve IDE support and make it crystal-clear what each argument must be.

When you add type hints (e.g. cfg: Config, model: nn.Module), your IDE can:

Auto-complete method and attribute names based on the expected types.

Spot errors early by flagging mismatched types as you type.

Provide tooltips and quick documentation pop-ups showing parameter types.

All of this makes writing and navigating code faster and less error-prone.

So, this part could be:



from typing import Type
import torch.nn as nn
from neuralngen.dataset.basedataset import BaseDataset

class BaseTrainer:
    def __init__(
        self,
        cfg: Config,
        model: nn.Module,
        dataset_class: Type[BaseDataset],
    ) -> None:
        """
        Trainer orchestrator for NeuralNgen.

        Args:
            cfg: Global configuration (paths, hyperparams, device, etc.).
            model: A PyTorch Module to be trained.
            dataset_class: Class that constructs train/val datasets.
        """
        …



2)
Original:


def _set_seed(self):
    random.seed(self.cfg.seed)
    np.random.seed(self.cfg.seed)
    torch.manual_seed(self.cfg.seed)
    torch.cuda.manual_seed_all(self.cfg.seed)

COMMENT:

Enabling deterministic CuDNN forces GPU ops to use fixed algorithms, giving bit-for-bit identical results across runs.


So, this part of the code could be like:

def _set_seed(self) -> None:
    seed = self.cfg.seed
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    # Force deterministic CuDNN (may slow training but guarantees repeatability)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False


In the second version:
torch.backends.cudnn.deterministic = True disables non-deterministic kernels.

torch.backends.cudnn.benchmark = False stops algorithm auto-tuning.

Combined with fixed random seeds, this guarantees reproducibility (at a small performance cost).

File: src/neuralngen/training/loss.py
Comment (FyavariFyavari) [inline src/neuralngen/training/loss.py:1]: There are few comments below:

1)

Original:


class ngenLoss(nn.Module):
    def __init__(
        self,
        cfg,
        distance_matrix=None,
        v_order=2,
    ):
        super().__init__()
        …


COMMENT:
Clear type hints and brief docstrings make code easier to read, enable smarter auto-completion and error checks in your editor, and catch misuse of functions early.

so, this part could be like:


from typing import Optional, Dict, Any
import torch
import torch.nn as nn

class ngenLoss(nn.Module):
    def __init__(
        self,
        cfg: Config,
        distance_matrix: Optional[torch.Tensor] = None,
        v_order: int = 2,
    ) -> None:
        """
        Combined loss:
          – Residual (MSE, RMSE, NSE)
          – Spatial variogram
          – Flow‐duration curve divergence

        Args:
            cfg: Hyperparameters (weights, loss type).
            distance_matrix: [D×D] for spatial weighting, or None.
            v_order: Semivariance power (default 2).
        """
        super().__init__()
        …


2)

original:


assert self.residual_loss.lower() in ["mse", "rmse", "nse"], \
    f"Invalid residualloss: {self.residual_loss}"
self.residual_loss = self.residual_loss.lower()


COMMENT:

Using ValueError with clear messaging is more Pythonic than assert, and remains active even if optimizations strip asserts.
So, the code could be like:


self.residual_loss = self.residual_loss.lower()
if self.residual_loss not in {"mse", "rmse", "nse"}:
    raise ValueError(f"Invalid residual_loss: '{self.residual_loss}'")




3)

Original:


loss = 0
losses = {}
…
loss += self.residual_weight * residual


COMMENT:

Starting with a Python 0 can trigger implicit conversions or device‐mismatch errors when mixed with GPU tensors. By using torch.zeros((), device=y.device), you create a tensor on the correct device from the start. This ensures all subsequent loss operations stay as tensors, preserving proper gradient tracking and avoiding type/device errors.

So, this part could be:


loss = torch.zeros((), device=y.device)
losses: Dict[str, torch.Tensor] = {}
…
loss = loss + self.residual_weight * residual  # ensures tensor arithmetic


4)

Variogram_loss

original:


def _variogram_loss(self, y_hat, y, mask):
    """
    Compute spatial variogram loss.
    Expects y_hat and y of shape [B, T, D=1]
    """
    # 1) extract a single hydro‐characteristic per sample
    method = "quantile"
    y_hat_flat = self.var_hyd_char(
        y_hat, method=method, random_quantile=True
    ).squeeze(-1)  # [B]
    y_flat = self.var_hyd_char(
        y, method=method, random_quantile=True
    ).squeeze(-1)      # [B]

    # 2) full B×B semivariance matrices via broadcasting
    diff_obs  = torch.abs(y_flat[:, None]   - y_flat[None, :])  ** self.v_order
    diff_pred = torch.abs(y_hat_flat[:, None] - y_hat_flat[None, :]) ** self.v_order

    # 3) squared‐difference matrix
    loss_matrix = (diff_obs - diff_pred) ** 2

    # 4) optional spatial weighting
    if self.distance_matrix is not None:
        weight = self.distance_matrix.to(y.device)  # [B×B]
        loss_matrix = loss_matrix * weight

    # 5) mean over all entries
    return loss_matrix.mean()





COMMENT:

the variogram_loss could be:



def _variogram_loss(self, y_hat, y, mask):
    """
    Compute spatial variogram loss using a full B×B semivariance matrix
    (zeros on the diagonal, symmetric).

    Expects y_hat and y of shape [B, T, D=1].
    """
    # 1) extract one summary per basin
    method = "quantile"
    y_hat_flat = self.var_hyd_char(
        y_hat, method=method, random_quantile=True
    ).squeeze(-1)  # [B]
    y_flat = self.var_hyd_char(
        y, method=method, random_quantile=True
    ).squeeze(-1)      # [B]

    B = y_flat.shape[0]

    # 2) reshape to [B,1] for pairwise distance
    y_obs  = y_flat.view(B, 1)
    y_pred = y_hat_flat.view(B, 1)

    # 3) compute condensed pairwise |yi - yj|^v_order
    obs_vec  = torch.pdist(y_obs,  p=self.v_order)  # length = B*(B-1)/2
    pred_vec = torch.pdist(y_pred, p=self.v_order)

    # 4) elementwise squared difference
    sq_diff = (obs_vec - pred_vec) ** 2

    # 5) build full B×B loss matrix (zeros on diag, symmetric)
    loss_matrix = torch.zeros((B, B), device=y.device)
    idx = torch.triu_indices(B, B, offset=1)  # indices for i<j
    loss_matrix[idx[0], idx[1]] = sq_diff     # fill upper triangle
    loss_matrix[idx[1], idx[0]] = sq_diff     # mirror to lower triangle

    # 6) optional spatial weighting
    if self.distance_matrix is not None:
        weight = self.distance_matrix.to(y.device)  # [B×B]
        loss_matrix = loss_matrix * weight

    # 7) aggregate final loss
    return loss_matrix.mean()
